src/source/utils.js

```diff
@@ -37,6 +78,99 @@ export function getS3Key(org, site, path) {
  * @return {string} the source bus path
  */
 export function getSourceKey(info) {
-  const { org, site, resourcePath: key } = info;
-  return getS3Key(org, site, key);
+  const { org, site, resourcePath } = info;
+  return getS3Key(org, site, resourcePath);
+}
+
+/**
+ * Validate the HTML message body stored in the request info.
+ *
+ * @param {import('../support/AdminContext').AdminContext} context context
+ * @param {import('../support/RequestInfo').RequestInfo} info request info
+ * @returns {Promise<Buffer>} body the message body as buffer
+ */
+export async function validateHtml(context, info) {
+  function validateHtmlError(message) {
+    const msg = `${message.message} - ${message.note}`;
+    if (ACCEPTABLE_HTML_ERRORS.includes(message.ruleId)) {
+      context.log.warn(`Ignoring HTML error: ${msg}`);
+      return;
+    }
+    throw new StatusCodeError(msg, 400);
+  }
+
+  const body = await info.buffer();
+
+  // TODO Check HTML size limit
+
+  fromHtml(body.toString(), {
+    onerror: validateHtmlError,
+  });
+  return body;
+}
+
+/**
+ * Validate the JSON message body stored in the request info.
+ *
+ * @param {import('../support/AdminContext').AdminContext} context context
+ * @param {import('../support/RequestInfo').RequestInfo} info request info
+ * @returns {Promise<Buffer>} body the message body as buffer
+ */
+export async function validateJson(context, info) {
+  const body = await info.buffer();
+
+  // TODO check JSON size limit
+
+  try {
+    JSON.parse(body);
+  } catch (e) {
+    throw new StatusCodeError(`Invalid JSON: ${e.message}`, 400);
+  }
+  return body;
+}
+
+/**
+ * Validate media body stored in the request info.
+ *
+ * @param {import('../support/AdminContext').AdminContext} context context
+ * @param {import('../support/RequestInfo').RequestInfo} info request info
+ * @param {string} mime media type
+ * @returns {Promise<Buffer>} body the message body as buffer
+ */
+export async function validateMedia(context, info, mime) {
+  const mediaType = MEDIA_TYPES.find((type) => type.mime === mime);
+  if (!mediaType) {
+    throw new StatusCodeError(`Unknown media type: ${mime}`, 400);
+  }
+  const body = await info.buffer();
+  try {
+    await mediaType.validate(context, info.resourcePath, body);
+  } catch (e) {
+    let msg = e.message;
+    if (msg.startsWith(PREVIEW_ERROR_PREFIX)) {
+      // Change the error message to not mention preview
+      msg = msg.replace(PREVIEW_ERROR_PREFIX, 'Media not accepted');
+    }
+    throw new StatusCodeError(msg, 400);
+  }
+  return body;
+}
+
+/**
+ * Validate the body stored in the request info.
+ *
+ * @param {import('../support/AdminContext').AdminContext} context context
+ * @param {import('../support/RequestInfo').RequestInfo} info request info
+ * @param {string} mime media type
+ * @returns {Promise<Buffer>} body the message body as buffer
+ */
+export async function validateUpload(context, info, mime) {
```

Comment:
somehow strange that `validateUpload` returns something. maybe rename it to `getValidPayload()` ? don't know.

Reply:
I agree, I've renamed it.